File: neet_code_150_cpp/graphs/redundant_connection.cpp
#include <iostream>
#include <vector>

// 模範解答
// Union-Findを使う
class SolutionAns {
    public:
        std::vector<int> findRedundantConnection(std::vector<std::vector<int>>& edges) {
            auto n = (int)edges.size();

            std::vector<int> parents;
            std::vector<int> ranks;

            for(int i = 0; i <= n; ++i) {
                parents.push_back(i);
                ranks.push_back(i);
            } 

            std::vector<int> result;
            for(auto& v: edges) {
                auto [n_1, n_2] = std::pair(v[0], v[1]);

                if(!doUnion(parents, ranks, n_1, n_2)) {
                    result = {n_1, n_2};
                    break;
                }
            }

            return result;
        }

    private:
        int doFind(std::vector<int>& parents, int n) {
            auto p = parents[n];
            while(p != parents[p]) {
                parents[p] = parents[parents[p]];
                p = parents[p];
            }

            return p;
        }

        bool doUnion(std::vector<int>& parents, std::vector<int>& ranks, int n_1, int n_2) {
            auto [p_1, p_2] = std::pair(doFind(parents, n_1), doFind(parents, n_2));

            if(p_1 == p_2) {
                return false;
            }

            if(ranks[p_1] > ranks[p_2]) {
                parents[p_2] = p_1;
                ranks[p_1] += ranks[p_2];
            } else {
                parents[p_1] = p_2;
                ranks[p_2] += ranks[p_1];
            }

            return true;
        }
};

// AC
// Union-Findの実装が少し違う
class SolutionRe {
    public:
        std::vector<int> findRedundantConnection(std::vector<std::vector<int>>& edges) {
            auto n = (int)edges.size();

            std::vector<int> parents(n + 1, -1);
            std::vector<int> size_graph(n + 1, 1);
            std::vector<int> result;

            for(auto& v: edges) {
                auto [n_1, n_2] = std::pair(v[0], v[1]);

                if(!unite(n_1, n_2, parents, size_graph)) {
                    result = {n_1, n_2};
                    break;
                }
            }

            return result;
        }

    private:
        // for Union-Find
        int root(int x, std::vector<int>& parents) {
            if(parents[x] == -1) {
                return x;
            } else {
                return parents[x] = root(parents[x], parents);
            }
        }

        bool unite(int x,
                int y,
                std::vector<int>& parent,
                std::vector<int>& size) {

            x = root(x, parent);
            y = root(y, parent);

            if(x == y) {
                return false;
            }

            if(size[x] < size[y]) {
                std::swap(x, y);
            }

            parent[y] = x;
            size[x] += size[y];

            return true;
        }
};

void show_result(const std::vector<int>& result) {
    for(auto& v: result){
        std::cout << v << " ";
    }
    std::cout << std::endl;
}

int main(void) {
    auto case_1 = std::vector{std::vector{1, 2}, std::vector{1, 3}, std::vector{2, 3}};
    // => [2, 3]
    auto case_2 = std::vector{std::vector{1, 2}, std::vector{2, 3}, std::vector{3, 4}, std::vector{1, 4}, std::vector{1, 5}};
    // => [1, 4]

    SolutionAns s_ans;
    auto res_1 = s_ans.findRedundantConnection(case_1);
    auto res_2 = s_ans.findRedundantConnection(case_2);

    show_result(res_1);
    show_result(res_2);

    SolutionRe s_re;
    auto res_1_re = s_re.findRedundantConnection(case_1);
    auto res_2_re = s_re.findRedundantConnection(case_2);

    show_result(res_1_re);
    show_result(res_2_re);
}

File: neet_code_150/src/graphs/redundant_connection.rs
use std::collections::{HashMap, HashSet};

struct Solution {}
impl Solution {
    pub fn find_redundant_connection(edges: Vec<Vec<i32>>) -> Vec<i32> {
        let mut set: HashSet<i32> = HashSet::new();
        let mut result = vec![];

        for edge in &edges {
            if set.contains(&edge[1]) {
                result.push(edge[0]);
                result.push(edge[1]);
            } else {
                set.insert(edge[1]);
            }
        }

        result
    }
}

// AC
// C++の模範解答より
// Union Findを使う
struct SolutionAnsCpp {}
impl SolutionAnsCpp {
    pub fn find_redundant_connection(edges: Vec<Vec<i32>>) -> Vec<i32> {
        let n = edges.len();

        let mut parents = vec![];
        let mut ranks = vec![];

        for i in 0..=(n as i32) {
            parents.push(i);
            ranks.push(i);
        }

        let mut result = vec![];
        for edg in edges {
            let (n_1, n_2) = (edg[0], edg[1]);

            if !Self::do_union(&mut parents, &mut ranks, n_1, n_2) {
                result = vec![n_1, n_2];
                break;
            }
        }

        result
    }

    // Union-Findのクエリの実装
    fn do_find(parents: &mut Vec<i32>, n: i32) -> i32 {
        let mut p = parents[n as usize];

        while p != parents[p as usize] {
            parents[p as usize] = parents[parents[p as usize] as usize];
            p = parents[p as usize];
        }

        p
    }

    fn do_union(parents: &mut Vec<i32>, ranks: &mut Vec<i32>, n_1: i32, n_2: i32) -> bool {
        let (p_1, p_2) = (Self::do_find(parents, n_1), Self::do_find(parents, n_2));

        if p_1 == p_2 {
            return false;
        }

        if ranks[p_1 as usize] > ranks[p_2 as usize] {
            parents[p_2 as usize] = p_1;
            ranks[p_1 as usize] += ranks[p_2 as usize];
        } else {
            parents[p_1 as usize] = p_2;
            ranks[p_2 as usize] += ranks[p_1 as usize];
        }

        true
    }
}

// AC
// Union-Findの実装が少し違う
struct SolutionRe {}
impl SolutionRe {
    pub fn find_redundant_connection(edges: Vec<Vec<i32>>) -> Vec<i32> {
        let n = edges.len();

        let mut parents = vec![-1; n + 1];
        let mut size_graph = vec![1; n + 1];
        let mut result = vec![];

        for v in edges {
            let (n_1, n_2) = (v[0], v[1]);

            if !Self::unite(n_1, n_2, &mut parents, &mut size_graph) {
                result = vec![n_1, n_2];
                break;
            }
        }

        result
    }

    fn root(x: i32, parents: &mut Vec<i32>) -> i32 {
        if parents[x as usize] == -1 {
            return x;
        }

        parents[x as usize] = Self::root(parents[x as usize], parents);
        return parents[x as usize];
    }

    fn unite(mut x: i32, mut y: i32, parents: &mut Vec<i32>, size_graph: &mut Vec<i32>) -> bool {
        x = Self::root(x, parents);
        y = Self::root(y, parents);

        if x == y {
            return false;
        }

        if size_graph[x as usize] < size_graph[x as usize] {
            std::mem::swap(&mut x, &mut y);
        }

        parents[y as usize] = x;
        size_graph[x as usize] += size_graph[y as usize];

        return true;
    }
}

fn main() {
    let case_1 = vec![vec![1, 2], vec![1, 3], vec![2, 3]];
    // => [2, 3]
    let case_2 = vec![vec![1, 2], vec![2, 3], vec![3, 4], vec![1, 4], vec![1, 5]];
    // => [1, 4]
    let case_3 = vec![vec![1, 3], vec![3, 4], vec![1, 5], vec![3, 5], vec![2, 3]];
    // => [3, 5]

    /*
    println!(
        "case_1: {:?}",
        Solution::find_redundant_connection(case_1.clone())
    );
    println!(
        "case_2: {:?}",
        Solution::find_redundant_connection(case_2.clone())
    );
    */

    println!(
        "case_1: {:?}",
        SolutionAnsCpp::find_redundant_connection(case_1.clone())
    );
    println!(
        "case_2: {:?}",
        SolutionAnsCpp::find_redundant_connection(case_2.clone())
    );

    println!(
        "case_1: {:?}",
        SolutionRe::find_redundant_connection(case_1.clone())
    );
    println!(
        "case_2: {:?}",
        SolutionRe::find_redundant_connection(case_2.clone())
    );
}
